c/misra/src/rules/RULE-1-5/CallToReallocWithSizeZero.ql

```diff
@@ -0,0 +1,25 @@
+/**
+ * @id c/misra/call-to-realloc-with-size-zero
+ * @name RULE-1-5: Disallowed size argument value equal to zero in call to realloc
+ * @description Invoking realloc with a size argument set to zero is implementation-defined behavior
+ *              and declared as an obsolete feature in C18.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/misra/id/rule-1-5
+ *       correctness
+ *       external/misra/c/2012/amendment3
+ *       external/misra/obligation/required
+ */
+
+import cpp
+import codingstandards.c.misra
+import semmle.code.cpp.rangeanalysis.new.RangeAnalysis
+
+from FunctionCall call, Expr arg
+where
+  not isExcluded(call, Language4Package::callToReallocWithSizeZeroQuery()) and
+  call.getTarget().hasGlobalOrStdName("realloc") and
+  arg = call.getArgument(1) and
+  upperBound(arg) = 0
```

Comment:
I can think of two different ways to write this rule:
 * Look for the use of `0` as a literal, that is then used in the realloc call.
 * Use range analysis, and report if `0` is a possible value that can be assigned.

The first is more precise - we're unlikely to get false positives - but may miss cases where a computed value is used in a `realloc` call which could be zero in some circumstances.

The second approach addresses that problem, but can cause false positives because our range is a single contiguous range and may not be able to exclude cases where 0 isn't possible.

In general, I think the range analysis approach is reasonable, although we want to confirm that it doesn't create significant false positives in common scenarios.

Your current implementation is a restricted version of the range analysis approach, because you force the upper bound to be zero. Using this approach, we would not find, e.g., this case:
```cpp
int n = some_bool ? 0 : 5;
realloc(t,n)
```
This is because the range of `n` is `0..5`, so the `upperBound` is `5`, even though `0` is set specifically.

I also don't think `upperBound` takes into account conversions. Given realloc takes `size_t` (i.e. an unsigned value), this type of scenario - which is morally equivalent to the above - would be reported:
```
int n = some_bool ? 0 : -5;
realloc(t,n)
```

I think my recommendations would be to try using `upperBound(arg.getConversion()) >= 0 and lowerBound(arg.getConversion()) <=0`, and review using MRVA to see how many FPs we get, and whether we can address them.

Reply:
Running MRVA, this may have something like a 20% ish false positive rate (90 results out of 512 total calls to `realloc()`).

I'm not sure how to address these false positives. It looks like most of them are allocating a fixed number of bytes extra per some event, and either that event parameter takes an `int` size instead of a `uint size`, and/or, range analysis might be predicting possible overflow.

Other false positives are the result of doing more complex computations (perhaps, multiplication, subtracting from space remaining, etc).

Given that `realloc()` is already generally banned...I'm not sure if it is common that customers run these misra suites on non-embedded code where `malloc()` is allowed & used....I'm not sure if we should err on the side of false positives here, or false negatives here, etc.

And of course, for codebases worried about compliance, which for some reason have decided to allow `malloc`/`free`/`realloc`, they should be able to insert checks that `size > 0` to better inform range analysis and suppress false positives.